### ISSUE TITLE
Port NBuilder to Dot Net Core

### ISSUE THREAD

**Issue by crmckenzie** [nbuilder/nbuilder#55]

**Comment by crmckenzie** [comment]
https://docs.microsoft.com/en-us/dotnet/articles/core/porting/

**Comment by ckolumbus** [comment]
I made a first attempt in porting, available in my [cloned repo](https://github.com/ckolumbus/nbuilder/tree/dncore). Library builds and a little test program works as well but I have not yet ported the tests. I was fighting with supporting the full .net framework parallel to .net core with one source dir and only an additional xproj file. 
According to my found resources it seem not to be possible to have both csproj & xproj project in one dir (clashes regarding project.json and not beeing able to "link" source file into two directories, [see issue here](https://github.com/dotnet/cli/issues/4022) ). Also it seems according to [this](https://blogs.msdn.microsoft.com/dotnet/2016/05/23/changes-to-project-json/) that the build structure will change in .net core also.
Any suggestions/ideas on how to organize the build tree under the given circumstances?

**Comment by crmckenzie** [comment]
I've done the same. What I've figured out is that I first need to upgrade to .NET 4.6, migrate to use .NET Core compatible libraries, and save the actual code conversion for the end. 

I've done the 4.6 upgrade on a branch, though I've apparently not pushed that branch up yet.

The next step will be to convert the RhinoMocks usage in the tests to NSubstitute which has a .NET Core implementation. The actual code in in NBuilder converts rather easily, but I'm not willing to do it without taking the working tests along for the ride.

When I get home tonight I'll push up my working branch. If you want to help me with the RhinoMocks => NSubstitute conversion I'd really appreciate it. From there it should be pretty easy.

As to your specific question regarding the new source tree--I'm going to create a new folder `/src` and put the .NET Core conversion in it--but only after it is essentially a copy-paste operation. My understanding is that with a single solution and set of source code I should be able to target multiple platforms. This will be an improvement over what we have now, though it will necessitate dropping support for .NET 4.0 and earlier. I'm okay with that. The world is moving on--so should we.

**Comment by PureKrome** [comment]
Nsubstitute over Moq?

![image](https://cloud.githubusercontent.com/assets/899878/21119716/f03fb7b6-c117-11e6-9026-df773c3bf5e3.png)

Also, Moq and xUnit are the two frameworks used in vNext stuff. 

So keeping with familiarity, maybe use those two instead?

**Comment by crmckenzie** [comment]
@PureKrome I have a strong preference for NSubstitute over Moq, and since I do most of the work on this project these days I'd prefer to use it. NUnit has a .NET Core implementation, so I don't see converting to xUnit as adding value.

I'm staying focused on "done." :)

Feel free to argue with me.

**Comment by PureKrome** [comment]
not going to argue at all :) reasons are perfectly justified 👍

**Comment by crmckenzie** [comment]
RhinoMocks is no more. Next will be the 2nd attempt at the .NET Core conversion.

**Comment by ckolumbus** [comment]
Regarding NUnit: dotnet-test-nunit only exists as v.3.4.0-beta3 and is not compatible with latest netcoreapp version. I'm not able to get the tests running with my update NBuilder code for .net core.

Options: 

1. wait for dotnet-test-nunit to upgrade and support latest .Netstandard (but there's a stall since 2 month)
2. make NBuilder .netstandard1.0 
3. switich to xUnit (which is against your wish above).

Ideas? Suggestions?

**Comment by ckolumbus** [comment]
I do have UTs running (not the functional ones yet) : checkout [my dncore branch](https://github.com/ckolumbus/nbuilder/tree/dncore)

go to Fizzware.NBuilder
* dotnet restore && dotnet build

go to Fizzware.NBuilder.Tests
* dotnet restore && dotnet build && dotnet test

its based on netstandard1.0 & netcoreapp1.0 that is also supported by dotnet-test-nunit.

**Comment by ckolumbus** [comment]
Some more information on netstandard vs netcoreapp wrt NUnit can be found in [this issue](https://github.com/nunit/dotnet-test-nunit/issues/42)

**Comment by lorddev** [comment]
@crmckenzie bump

**Comment by crmckenzie** [comment]
I've got another PR (61) doing similar work. When MS introduced .NET Standard and reverted to .csproj files I became concerned that we'd need to do another revision of this work to keep things current. To my knowledge, there's no current version of .NET Standard that works with both 4.6/7 and dotnetcore, but that is coming. My feeling is that we should concentrate on building a .NET Standard version of NBuilder and abandon using 3-4 solutions to recompile the same code for different platforms.

Thoughts?

**Comment by PureKrome** [comment]
I thought `.NET Standard 1.3` is for .NET 4.6+ _and_ .NET Core?

ref: [.NET Standard Platform Support Matrix](https://docs.microsoft.com/en-us/dotnet/articles/standard/library)

That said, I would assume it should be easier now to support more platforms under a single solution if you upgrade the project to using .NET Core vs multi-sln+.Net Old.

**Comment by lorddev** [comment]
.NET Standard 1.6 is compatible with 4.7. .NET Standard 1.1 is compatible with 4.5.2. .NET Standard 2.0 will be backwards compatible with basically everything.

**Comment by crmckenzie** [comment]
Okay--let's do .NET Standard 1.3 then.

**Comment by PureKrome** [comment]
very very, warm and polite _bump_ on this?

**Comment by ckolumbus** [comment]
Hi all,
a carful question here: why targeting the latest and greatest runtimes when it doesn't seem to be necessary. At least the main lib is fully capable of being build against net461.. which I actually need because 4.7 is not installed in my environment and I rebuild it by just changing the frameworks target in csproj. 

I'm also not sure whether netstandard1.6 is really necessary either. The following tool might be helpful, although I've not yet tested it myself. https://docs.microsoft.com/en-us/dotnet/standard/portability-analyzer
IMHO we should try to target the minimally necessary API surface/runtime that is absolutely necessary to make the lib available to as many environments as possible.
Just my 2c. Opinions?

PS: if the test suites need 4.7 that would be fine because it's for building only, but the lib itself could at least be targeting lower versions.

**Comment by crmckenzie** [comment]
I'm close to having a working build for 3.5, 4.0, 4.7, and Standard 1.6. The compilation part works--I just have to get the build scripts all straightened out.

**Comment by ckolumbus** [comment]
@crmckenzie great to hear this 👍  but do we really have platform specific code in the 4.7/1.6 arena that needs such a big API surface? Just to make myself maybe a little clearer: I'm only talking about the build for netstandard. I understand the specific  3.0 & 4.0 versions, but for eveything else we could check what the lowest netstandard version is that we need for the lib. Then we could support much more than with net47/netstand1.6 as this creates a gap for all netstandard based projects that do not want to target the full framework and support lower netstandard versions. (== everything with support for netstandard1.1 - 1.5).

All this holds only if  I understand the [matrix here](https://docs.microsoft.com/de-de/dotnet/standard/net-standard)  and the info about standards/targets and so on[here](https://blogs.msdn.microsoft.com/cesardelatorre/2016/06/28/running-net-core-apps-on-multiple-frameworks-and-what-the-target-framework-monikers-tfms-are-about/) and [here](https://github.com/dotnet/standard/blob/master/docs/faq.md) correctly. 

With some luck I find some time to check the code myself so that I can help with concrete PRs and not just talking :-) But at least I want to know whether I made myself clear and what you think about this topic.

**Comment by PureKrome** [comment]
> 3.5, 4.0, 4.7, and Standard 1.6

I've not ran the port analyser on this either ... but .... if it's all pretty common stuff, then it could target

- 3.5 (old / back compat)
- 4.0 (again, old / back compat)
- 4.5 (again, old and back compat)
- 4.6 == .NET Standard 1.3  (so that's 4.6+)

in theory, you could do this too..
- 3.5
- 4.0
- .NET Standard 1.0 (which is 4.5+)

**Comment by lorddev** [comment]
@ckolumbus See PR #61 -- basically reflection stuff and trace logging had to be adjusted for platform-specific reasons. It's possible these APIs have been added back in for netstandard 2 but I haven't checked yet.

**Comment by PureKrome** [comment]
Is there anything we (the community) can do right now, to help?

**Comment by PureKrome** [comment]
Oh also - speaking of frameworks once more ... does it have to target those low ones? could it possibly just target .NETStandard 1.4 and be done with it? The other frameworks can be packages up on NuGet that are obsolete or somehow maker as the most recent ?

**Comment by lorddev** [comment]
@PureKrome I suppose testing/using my fork could help, in order to have more eyes on it. Then we might be able to persuade the Chrisses to move forward. Another thing I've been thinking about is whether we can remove some of the duplicate sln files like you said.

**Comment by crmckenzie** [comment]
@ckolumbus "but do we really have platform specific code in the 4.7/1.6"
Unknown. One of the challenges I have is that the dev tools for dotnet core 1.0 and 1.1 do not play nicely side-by-side. My thinking was to err on the side of dotnet core 1.1 since it's newer and presumably represents the direction that MS is moving.

As to the progress on this issue: it's mostly done. I just need to find 4-6 uninterrupted hours to make the build scripts function again. I think I know _how_ to do it but finding the time to sit down with it has been difficult lately for life reasons. Perhaps I'll just push it up in its broken state and one of you can help me with it if you have the time.

**Comment by PureKrome** [comment]
> Perhaps I'll just push it up in its broken state and one of you can help me with it if you have the time.

When you say "_push it up_" I hope you can do this if it's all to a separate branch so then we can pull it down, review, tinker, play and hopefully share the love/load and get a fix up.

I've done a heap of work getting things to work on AppVeyor with powershell ... but my experience with custom build tools like cake, etc is low.

I'm _very_ eager to help because this project is one of my major blockers for migrating my existing stuff over to Core.

**Comment by PureKrome** [comment]
@lorddev Interesting you suggested reviewing #61 .. cause that's exactly what I did yesterday. First thing I did was clone this repo. Then switch to #61 then downgraded the framework from NS1.6 to NS1.4 ... deleted all the test projects .. then built .. and boom :(

This reflection stuff is killing me.

Any chance you could try and get the project to work on NS1.4 ?? 

When I do file->new->Standard Library project, it defaults to 1.4 (which is why i've been starting out targetting 1.4). But of course ...  I was going to see how low I could go.

**Comment by ckolumbus** [comment]
@crmckenzie just push it and I'll give it a try!

**Comment by lorddev** [comment]
@PureKrome @crmckenzie per https://stackoverflow.com/a/40685046/16454 BindingFlags did not exist in 1.4 so I will have to reference another library for it to build.

**Comment by lorddev** [comment]
@crmckenzie @PureKrome I'm sorry. It seems the reflection APIs in 1.4 are simply incompatible. Please consider using 1.6.

**Comment by crmckenzie** [comment]
The conversion to .NET Standard 1.6 is complete. The code compiles. The tests pass. What's left is the scripts used to build the package on AppVeyor. I'll try to push that up tonight so others can help me with them.

**Comment by PureKrome** [comment]
I have the day off so the sooner you push it up, the sooner I can look into it. 👍

**Comment by crmckenzie** [comment]
I've pushed to feature/55.  Here's what I think needs to happen:
1) The .NET Standard Solution needs .NET Standard projects for the tests. They currently build against .NET 4.7 which seems to confuse the NUnit test runner during the build/test process.
2) Invoke-EmulateAppVeyor is the script I use to build the nuget package with the various compilled versions of NBuilder.

My goal is to produce a nuget package with binaries for .NET 4.0, 4.7 and .NET Standard 1.6. Once Invoke-EmulateAppVeyor does that I'll put nbuilder through the release process.

If anyone makes progress on this, let me know. Otherwise I'll get to it as soon as I can.

**Comment by PureKrome** [comment]
Thanks for pushing this up!

**Comment by PureKrome** [comment]
@crmckenzie Yo - so I've started playing with your branch.

>The .NET Standard Solution needs .NET Standard projects for the tests

Okay. So i migrated `FizzWare.NBuilder.Tests` to .NS1.6. I'm a noob at this so it took a bit of meddling to get it to compile. Next issue was getting nunit to run at all. Doesn't get discovered in Resharper but it does get discovered in Test Explorer. *415* tests passed.

Next, I'm trying now with `FizzWare.FunctionalTests.Model` project. But this has a dependency on `EntityFramework 6.1.3`. Which ... is a problem. This don't exist/work under a .NET Core app :(

So - do we need the tests to be running as .NET Core? Or can we maybe run then as .NET 4.7 framework but under a .NET Core prj (yes, this is so confusing for me still).

What _issues_ did you have?

**Comment by crmckenzie** [comment]
Quick note: Be careful with "migrate"--we should build new test projects pointing at the same cs files so that the .NET 4 solution is unaffected by this work.

* Re: NUnit - Nunit works for me in VS, but it took some finagling. Where I left off is that it doesn't work from the command-line which I believe is the result of framework confusion.

* Re: EntityFramework - I don't have any easy answers for this. I was _hoping_ that we could just port the projects to .NET Standard and be done with it. It appears that Microsoft's framework churn has made that goal less viable. Worst case we can drop the integration tests from the .NET Standard solution until we have .NET Standard testrunner support.

* "do we need the tests to be running as .NET Core?" - Yes if possible. I've already uncovered and fixed some defects in NBuilder caused by subtle changes in framework implementations between .NET Standard and .NET 4.0. Without the tests we'd be shooting in the dark.

And in case I forget to mention it--thanks for taking a look at this.

**Comment by lorddev** [comment]
@crmckenzie tests are already broken. I created another issue: #62

**Comment by PureKrome** [comment]
>Quick note: Be careful with "migrate"--we should build new test projects pointing at the same cs files so that the .NET 4 solution is unaffected by this work.

Yep. that's exactly what I did => `Add new Project -> Standard Lib 1.6`. Then copied the old files over to the new project.

> Where I left off is that it doesn't work from the command-line which I believe is the result of framework confusion.

Ah! CLI. ok. It's been ... maybe a decade since i used NUnit last so all this is new to me (i.e. learning the new _tools_ to use). So I'll see if I can get something working there/then.

> EntityFramework

I'll try dropping it, in the mean time. Baby steps, etc.

> Without the tests we'd be shooting in the dark.

Totally agreed. I was more checking if the tests had to target a particular framework. .NET core 1.1 or .NET Framework 4.7 all under the new csproj format of a the new CORE project template, etc. Yeah - so confusing. Naming is hard. 😢 

>And in case I forget to mention it--thanks for taking a look at this.

Pleasure :) Full disclosure: my reasons are purely selfish though (even though I do love doing OSS of course!) This library is literally one of the only ones now, that are stopping me from going 100% .NET core. I think i have one other.. i think. But yeah :)

Many hands make light work 👍 

(I hope my tinkering can end up becoming helpful and contributable. 😊 )

**Comment by lorddev** [comment]
`HasMany().WithMany()` doesn't work in EF 7. Otherwise I was able to finagle things. But it's kind of a blocker.

**Comment by crmckenzie** [comment]
Is there an EF7 equivalent? We could control that with preprocessor directives.

**Comment by PureKrome** [comment]
Maybe drop EF and go Dapper? (no small feat :( )

**Comment by crmckenzie** [comment]
@PureKrome I think it's easier than that. I'm looking into going in-memory. We're not testing Entity Framework--we're testing NBuilder's integration with some persistence mechanism. 

BTW, NUnit published a new version of their test adapter. I'm looking into going .NET Standard across the board. I'll let you know if I get anywhere.

**Comment by PureKrome** [comment]
Thanks heaps @crmckenzie - I totally know I'm sounding like a pushy, random OSS asshat so I totally appreciate volunteers finding time and trying to get things done in OSS projects, etc.

**Comment by crmckenzie** [comment]
After a couple of hours of investigation last night, I think things need to happen in a specific sequence at this point.

1. Remove EntityFramework from the FunctionalTests project. Instead use an in-memory implementation of BaseRepository<T>. This has the sideeffect of removing the "ProductCategories" table from the tests and associated assertions as that table would be created by an ORM when talking to a relational database. **I've already done this on my branch but it's comingled with other work.** I'm going to revert, redo, and push the changes to this branch.
2. Consolidate FunctionalTests and the standard Unit Tests into the same project as there's no longer anything that requires out-of-process resources. This has the benefit of reducing the number of test projects that have to be maintained.
3. As there are now only 2 projects, convert them both to .NET Standard 1.6
4. Port the tests to a testing framework that supports .NET Standard 1.6. We can keep the reference to NUnit to maintain the Assertions as they are currently coded but we need a test runner that will actually work. This framework still needs to be identified but I'm leaning toward MSTest if it works as it was basically modeled on NUnit so the port should be minimal.

**Constraints**
1. I do not want to maintain 2 copies of the source files for the 4.5 and .NET Standard solutions. Any code differences should be resolved via preprocessor directories (e.g. `#if NETSTANDARD1_6`) but .NET Standard projects and original projects should reference the same source files.
2. The ultimate goal is that Invoke-EmulateAppVeyor.ps1 compiles and tests the code and generates a nuget package with net35, net45, and netstandard16 versions of `nbuilder.dll`.
3. It's more important that the tests function during the AppVeyor process than in the Visual Studio or ReSharper Test Runners.

**Comment by crmckenzie** [comment]
The feature/55 branch is now ready for a testing framework port for the .NET Standard projects. I tried MS Test and got nowhere.

**Comment by PureKrome** [comment]
@crmckenzie can we please consider `xUnit` for the testing framework please? That's what' being used by the MS team for their repo's, etc.

ref: [Kestrel HTTP Server](https://github.com/aspnet/KestrelHttpServer/blob/dev/test/Kestrel.Core.Tests/AddressBinderTests.cs)
ref: [MVC]9https://github.com/aspnet/Mvc/blob/dev/test/Microsoft.AspNetCore.Mvc.Core.Test/AcceptedAtActionResultTests.cs)
ref: [EntityFramework](https://github.com/aspnet/EntityFramework/blob/dev/test/EFCore.InMemory.FunctionalTests/CustomValueGeneratorTest.cs)

**Comment by lorddev** [comment]
@PureKrome They've declined previous requests to use xunit. But tests are working for me with nunit... just not the previously-mentioned one.

**Comment by lorddev** [comment]
@PureKrome I think you're right. I've tried for a couple hours to get nunit working with .net core and I can't do it.

**Comment by PureKrome** [comment]
>  They've declined previous requests to use xunit.

Yep - and I agree with their rational for sticking with the current libraries. But now things have changed .. as in .. the older libraries are not working as easily with the newer ways.

So maybe this might be the catalyst for change ... ??

**Comment by crmckenzie** [comment]
I have no principled opposition to using xUnit other than that changing a test-framework is a non-trivial operation and must be justified. So far that hasn't been the case. 

That said, I've been unable to get _any_ testing framework (including xUnit) to work against a .NET Standard project. I have a prototype solution that I'm working on to prove that this or that testing framework will function but I haven't pushed it to github yet. I'm still exploring possible solutions. Show me a working solution where the assembly under test is .NET Standard and I'll start the port myself.

**Comment by PureKrome** [comment]
It's mon morning here. I'll see if I can get xUnit to work with the project at lunch or tonight. I've done a number of Core/NetStandard stuff with xunit. can confirm it _does_ work :)

**Comment by PureKrome** [comment]
>  Show me a working solution where the assembly under test is .NET Standard and I'll start the port myself.

Refer to the links above for the official MS Repo's - they are all using xUnit + moq.

**Comment by crmckenzie** [comment]
I've looked at that stuff and as I said I've tried to make it work but so far to no avail. I was able to get xunit to work with ReSharper, but not the command-line where it matters for the CI build.

**Comment by PureKrome** [comment]
@crmckenzie 


![image](https://user-images.githubusercontent.com/899878/28761250-ab0b2610-75f0-11e7-8986-48e17d3cdf40.png)

this was just for the `DatesTests.cs` class...

so far so good.
i'll try and do the entire `Unit` folder ... the move up to the `Integration` folder....

is this what you mean by that you couldn't get the tests running on the CLI ... cause I did straight away .. ???

or am I missing something/misunderstood something?


BTW - now that I can run the unit test project from the CLI I can easily do this in an `AppVeyor.yml` file.

This is also just targetting a .NET core app (which is the .NetStandard 1.6 library), not another framework like 4.5 or 3.5 ...

**Comment by PureKrome** [comment]
Ok - some more info for everyone.

First, DISCLAIMER: I don't have .NET 3.5 installed on my machine. _I also refuse to install it_ (please, don't hate .. just my personal reasons). So these steps EXCLUDE 3.5 but it can be easily added in if you have it installed 👍 

---

### Clean baseline.
1. Clone new, this repo.
2. switch `feature/55`. Pull latest upstream, just to be sure...

### First, get Multi-targetting (for the main project => `FizzWare.NBuilder`).
1. Grab/read some references. [MS Docs](https://docs.microsoft.com/en-us/dotnet/core/porting/project-structure) and [Rick Strahl](https://weblog.west-wind.com/posts/2017/Jun/22/MultiTargeting-and-Porting-a-NET-Library-to-NET-Core-20). Yep, there's more ref's but I found these two a great starting point amoungst a lot of the noise out there. Plus, they are recent (which is important in this time of quick influx/changes/turnaround).
2. Create new `src` and `tests` folders. This is so I don't mess with existing code (which is basically working) and starts to bring some _consistency_ to the project structure making it _easier_ and _less scary_ for people to quickly grok the setup and feel comfortable looking a project code.
3. Create a new blank `.sln` and stick this in the root. (ref to point (2) just above).
4. Copy the current .NET Standard project to `src\FizzWare.NBuilder` . IMPORTANT: Note there's no FRAMEWORK moniker post-pending. This means i'm following the suggested _first_ recommendation from the MS docs regarding "_Organizing your project to support .NET Framework and .NET Core_".
5. Now - the tricky bit. _Manually_ editing the new `src\FizzWare.NBuilder\FizzWare.NBuilder.csproj` file to handle *multi-targetting*.

here's a snippet:

```
  <PropertyGroup>
    <TargetFrameworks>netstandard1.6;net45</TargetFrameworks>
    <AssemblyName>FizzWare.NBuilder</AssemblyName>
    <RootNamespace>FizzWare.NBuilder</RootNamespace>
    <Description />
    <BaseIntermediateOutputPath>obj-NetStandard16</BaseIntermediateOutputPath>
  </PropertyGroup>

  <PropertyGroup Condition=" '$(TargetFramework)' == 'netstandard1.6'">
    <DefineConstants>NETCORE;NETSTANDARD;NETSTANDARD1_6</DefineConstants>
  </PropertyGroup>

  <PropertyGroup Condition=" '$(TargetFramework)' == 'net45'">
    <DefineConstants>NET45;NETFULL</DefineConstants>
  </PropertyGroup>
```

6. Refactor some time code which is having some crap fights between 1.6 and 4.5. using `#if NETSTANDARD1_6` etc solved this:.

e.g.

```
#if NETSTANDARD1_6
            var returnType = handler.GetMethodInfo().ReturnType;
#else
            var returnType = handler.Method.ReturnType;
#endif
```

So far so good! how about compiling, etc?

Well, lets go back to this bit of code...

`<TargetFrameworks>netstandard1.6;net45</TargetFrameworks>`

This means *Visual Studio* will use the _first_ framework in this list and compile everything against that.
So right now, VS is in .NET Standard 1.6 mode with my solution open.

Remember this next bit?
```
  <PropertyGroup Condition=" '$(TargetFramework)' == 'netstandard1.6'">
    <DefineConstants>NETCORE;NETSTANDARD;NETSTANDARD1_6</DefineConstants>
  </PropertyGroup>
```

Because I'm in 1.6 mode (that's the first framework listed in that list..) ... the *constants* are setup for 1.6 mode. Want proof? Check out the `properties` of the project...

![image](https://user-images.githubusercontent.com/899878/28762894-5a400482-75fd-11e7-9dd4-4bb1f7b102d4.png)


How do I now test 4.5 mode? Change the order of the frameworks in that one special line ... close solution .. open solution and it's not in 4.5 mode....

e.g. `<TargetFrameworks>net45;netstandard1.6</TargetFrameworks>`

### Making sure multiple dll's are created.
1. Drop down to console/CLI mode for this.
2. `cd` to the root folder of this project. For me it's `C:\projects\temp\nBuilder`.
3. `dotnet build -c Release`  <-- this looks for the `FizzWare.NBuilder.sln` i created (read back a few steps, for when I did this). Inside my sln has (so far) just one `.csproj` which is targetting 1.6 and 4.5. So ... it builds both :)

![image](https://user-images.githubusercontent.com/899878/28762993-2ab1ac1a-75fe-11e7-8108-60610096e449.png)

### Finally - multi-targetting testing
*This is the kicker* . Test projects are just like another form of host/application. Web apps are hosts. Console apps are hosts. So are tests. 
- TESTS ARE NOT .NET STANDARD PROJECTS.
- TESTS should be a .NET CORE APP (or full framework if u're still in that arena).
- If you're doing a .NET CORE APP then u can multi-target the tests :)

Most people (like me at first) think a Test Project is a .NET Standard app. Sad trumbone. It's not a library. It has to run, etc. It's not like the old days.

so .. with this powerful new information ...

1. Create the new test project => `\tests\FizzWare.NBuilder.Tests` <-- again, not there's no framework post-pended.
2. Update the `xUnit` packages to the _latest pre-release_. I don't usually use pre-releases but I had some trouble on the current stables (at the time of this post -> 2.2.0) and had instant luck with the pre's.

```
    <PackageReference Include="xunit" Version="2.3.0-beta3-build3705" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.0-beta3-build3705" />
```

![image](https://user-images.githubusercontent.com/899878/28763482-c9d0a38e-7601-11e7-8fe8-01529c0c8088.png)


3. Copy _all the files_ from the existing test project (which is incorrectly a .NS 1.6 project) to this new project.
4. Added the `Shouldly` nuget package.
5. Removed all the files except the first one to make sure I can get the tests passing..
6. Replaced NUnit/Assert code with xUnit + Shouldly code with 3x replace-all-in-file commands. Yay!
7. 💰 💰 💰 💰 💰 
8. 16 / 16 tests all pass, for both frameworks.

![image](https://user-images.githubusercontent.com/899878/28763590-94483e6a-7602-11e7-9899-1c1a88b45783.png)


So that means .. can I finish the rest?

... _nearly 2 hours later_ ...

OK - done.

I've update all the files in the Test project to use `xUnit` and `Shouldly` (tests are now more fluent and nicer to read). Most common update issue was how NUnit and xUnit handle their setup's differently. Just too labour to refactor.

All the tests but a few pass. So this is not so much an issue of the port, but an issue with the original tests:
- xUnit is now _really_ running parallel. So there's some bad tests that share state, etc. Running the tests by themselves, they pass. In a group ... they fight each other. Answer: Fix the test logic.
- Some error about Persistence setup: `FizzWare.NBuilder.PersistenceMethodNotFoundException : No persistence create method set up for Category. Add one using BuilderSettings.SetPersistenceCreateMethod()`. I have no idea what this means so i'll leave this to people who are smarter than me.

### Can we create nuget packages for multiple frameworks?
Yes we can. 👍 
1. Update the project pacakge meta data in VS (or notepad, etc)
2. `dotnet pack`

![image](https://user-images.githubusercontent.com/899878/28765760-b914cf8a-760f-11e7-831b-a6cef36c7d05.png)

of course, I didn't specify `-c Release` and the version number etc ... but this can all be done on the command line, etc.

----

So @crmckenzie - how does this look?

**Comment by crmckenzie** [comment]
@PureKrome this is brilliant. I'm going to pull your src and tests folders and re-orient Invoke-EmulateAppVeyor.ps1 around your layout. Once I can reproduce your results, I'll start working on the tests.

**Comment by PureKrome** [comment]
![](https://media.giphy.com/media/JIX9t2j0ZTN9S/giphy.gif)

![](https://media.giphy.com/media/yoJC2GnSClbPOkV0eA/giphy.gif)

**Comment by pjmagee** [comment]
@PureKrome  I have admired your gif usage since you turned up to the NBuilder project. You always make me smile with your reaction gifs and input to this project and I'm happy this has been given a new owner and NBuilder continues on from Gareth. 👍  to your everyones efforts so far :D I always read all the NBuilder updates when I'm at work.

**Comment by crmckenzie** [comment]
Alright @PureKrome you've convinced me. I'm going to retrace your steps using `git move` to preserve the file history. Thank you for leaving such a well-documented breadcrumb trail!

**Comment by crmckenzie** [comment]
Okay I think this feature is done. @PureKrome would you like to review feature/55 and see if I missed anything? In the meantime I'm going to check for any quick bug-fixes that should make it into the 5.0.x release and start working on those.

**Comment by PureKrome** [comment]
@crmckenzie did you get those erroring unit tests, to work / pass?

**Comment by crmckenzie** [comment]
Oh yeah. The port away from Entity Framework was initially implemented as a static in-memory collection which of course doesn't play nice with parallelized tests. I did some test-refactoring and am now getting consistent green.

**Comment by PureKrome** [comment]
Awesome! I'll check it out tonight and report back, etc.

Thanks!!

**Comment by PureKrome** [comment]
`.vs` folders shouldn't be checked in.

like this one in `\src\vs` -> https://github.com/nbuilder/nbuilder/tree/feature/55/src/.vs/config

Also, I don't think you need the `Properties\AssemblyInfo.cs` file any more? And the `CommonAssemblyInfo.cs` ??

Also, the `NBuilder.snk` exists in the `src` and `src\FizzWare.NBuilder` folders. Duplicates?

The rest Looks Good To Me. AWESOME!

**Comment by crmckenzie** [comment]
Code ported, tests pass, AppVeyor build succeeds. Merging back to develop. This work will be in the next release (coming soon).

**Comment by crmckenzie** [comment]
The beta package is up on nuget.org. @PureKrome , wanna pull it into a dotnet core project and verify the package is good?

https://www.nuget.org/packages/NBuilder/5.0.0-beta-135

I'm planning to release this weekend.

**Comment by PureKrome** [comment]
I _might_ not get a chance in the next 48 hours. I'm literally about to head off to the snow for 2 days.

I will have a fair bit of time in a car (to snow) or bus (from snow) so I might try and test away.

VERY EXCITING!

![](https://31.media.tumblr.com/43e63461d1e3f22a49b18dbf15227a1d/tumblr_inline_n3t10oQfIh1solpjm.gif)

**Comment by lorddev** [comment]
Going to switch all my projects that implemented a local dotnet standard fork of NBuilder to use the nuget one. I'll let you know how it goes!

**Comment by lorddev** [comment]
Works great!

**Comment by PureKrome** [comment]
Yep - works!!
